pydantic/_internal/_typing_extra.py

```diff
@@ -265,24 +265,38 @@ def eval_type_backport(
                 value, globalns, localns
             )
     except TypeError as e:
-        if not (isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)):
-            raise
-        try:
-            from eval_type_backport import eval_type_backport
-        except ImportError:
+        if isinstance(value, typing.ForwardRef):
+            if is_backport_fixable_error(e):
+                try:
+                    from eval_type_backport import eval_type_backport
+                except ImportError:
+                    raise TypeError(
+                        f'Unable to evaluate type annotation {value.__forward_arg__!r}. If you are making use '
+                        'of the new typing syntax (unions using `|` since Python 3.10 or builtins subscripting '
+                        'since Python 3.9), you should either replace the use of new syntax with the existing '
+                        '`typing` constructs or install the `eval_type_backport` package. Otherwise, it might be '
+                        "that the type being used isn't subscriptable."
+                    ) from e
+
+                return eval_type_backport(value, globalns, localns, try_default=False)
+
             raise TypeError(
-                f'You have a type annotation {value.__forward_arg__!r} '
-                f'which makes use of newer typing features than are supported in your version of Python. '
-                f'To handle this error, you should either remove the use of new syntax '
-                f'or install the `eval_type_backport` package.'
+                f'Unable to evaluate type annotation {value.__forward_arg__!r}. '
+                "It might be that the type being used isn't subscriptable."
             ) from e
 
-        return eval_type_backport(value, globalns, localns, try_default=False)
+        raise e
 
 
 def is_backport_fixable_error(e: TypeError) -> bool:
     msg = str(e)
-    return msg.startswith('unsupported operand type(s) for |: ') or "' object is not subscriptable" in msg
+
+    return (
+        sys.version_info <= (3, 8)
+        and msg.startswith('unsupported operand type(s) for |: ')
+        or sys.version_info <= (3, 9)
+        and "' object is not subscriptable" in msg
```

Comment:
These versions look backwards, and that `<=` looks wrong.

```
➜  ~ python3.9                                        
Python 3.9.19 (main, Aug  2 2024, 16:30:42) 
[Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> int | str
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'type' and 'type'
>>> list[int]
list[int]
>>> import sys
>>> sys.version_info
sys.version_info(major=3, minor=9, micro=19, releaselevel='final', serial=0)
>>> tuple(sys.version_info)
(3, 9, 19, 'final', 0)
>>> sys.version_info <= (3, 9)
False
>>> (3, 8, 0) <= (3, 8)
False
```

Reply:
Indeed comparing with `<=` is confusing.

As a side note, I've raised https://github.com/microsoft/pyright/discussions/8648.